% {{{ Preamble


%% BioMed_Central_Tex_Template_v1.05
%%                                      %
%  bmc_article.tex            ver: 1.05 %
%                                       %


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                     %%
%%  LaTeX template for BioMed Central  %%
%%     journal article submissions     %%
%%                                     %%
%%         <27 January 2006>           %%
%%                                     %%
%%                                     %%
%% Uses:                               %%
%% cite.sty, url.sty, bmc_article.cls  %%
%% ifthen.sty. multicol.sty		       %%
%%									   %%
%%                                     %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                                                 %%	
%% For instructions on how to fill out this Tex template           %%
%% document please refer to Readme.pdf and the instructions for    %%
%% authors page on the biomed central website                      %%
%% http://www.biomedcentral.com/info/authors/                      %%
%%                                                                 %%
%% Please do not use \input{...} to include other tex files.       %%
%% Submit your LaTeX manuscript as one .tex document.              %%
%%                                                                 %%
%% All additional figures and files should be attached             %%
%% separately and not embedded in the \TeX\ document itself.       %%
%%                                                                 %%
%% BioMed Central currently use the MikTex distribution of         %%
%% TeX for Windows) of TeX and LaTeX.  This is available from      %%
%% http://www.miktex.org                                           %%
%%                                                                 %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


\NeedsTeXFormat{LaTeX2e}[1995/12/01]
\documentclass[10pt]{bmc_article}    



% Load packages
\usepackage{cite} % Make references as [1-4], not [1,2,3,4]
\usepackage{url}  % Formatting web addresses  
\usepackage{ifthen}  % Conditional 
\usepackage{multicol}   %Columns
\usepackage[utf8]{inputenc} %unicode support
%\usepackage[applemac]{inputenc} %applemac support if unicode package fails
%\usepackage[latin1]{inputenc} %UNIX support if unicode package fails
\urlstyle{rm}
 
 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%	
%%                                             %%
%%  If you wish to display your graphics for   %%
%%  your own use using includegraphic or       %%
%%  includegraphics, then comment out the      %%
%%  following two lines of code.               %%   
%%  NB: These line *must* be included when     %%
%%  submitting to BMC.                         %% 
%%  All figure files must be submitted as      %%
%%  separate graphics through the BMC          %%
%%  submission process, not included in the    %% 
%%  submitted article.                         %% 
%%                                             %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%                     

\def\includegraphic{}
\def\includegraphics{}

\setlength{\topmargin}{0.0cm}
\setlength{\textheight}{21.5cm}
\setlength{\oddsidemargin}{0cm} 
\setlength{\textwidth}{16.5cm}
\setlength{\columnsep}{0.6cm}

\newboolean{publ}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                              %%
%% You may change the following style settings  %%
%% Should you wish to format your article       %%
%% in a publication style for printing out and  %%
%% sharing with colleagues, but ensure that     %%
%% before submitting to BMC that the style is   %%
%% returned to the Review style setting.        %%
%%                                              %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%Review style settings
\newenvironment{bmcformat}{\begin{raggedright}\baselineskip20pt\sloppy\setboolean{publ}{false}}{\end{raggedright}\baselineskip20pt\sloppy}

%Publication style settings
%\newenvironment{bmcformat}{\fussy\setboolean{publ}{true}}{\fussy}

% Scaffolder url
\urldef{\scaffolder}\url{http://www.michaelbarton.me.uk/projects/scaffolder/}

% Begin ...
\begin{document}
\begin{bmcformat}
% }}}
% {{{ Title

\title{Scaffolder - Software for Microbial Genome Scaffolding.}

\author{
  Michael D Barton$^{1}$%
  \email{Michael D Barton - mail@michaelbarton.me.uk}%
\and
  Hazel A Barton\correspondingauthor$^1$%
  \email{Hazel A Barton\correspondingauthor - bartonh@nku.edu}%
      }

\address{\iid(1) Department of Biological Sciences, Northern Kentucky%
University, Nunn Drive, Highland Heights, KY 41076 }%

\maketitle
%}}}
% {{{ Abstract

\begin{abstract}

  \paragraph*{Background:} Assembly of short read sequencing data can result in
  a fragmented series of contigs. Therefore a common step in a genome project
  is to join neighbouring sequence regions together and fill gaps using insert
  data generated through PCR. This scaffolding step, however, is non-trivial
  and requires manually editing large blocks of nucleotide sequence. Joining
  sequence contigs together also hides the source of each region in the final
  genome sequence. Taken together, these considerations may make reproducing or
  editing an existing genome build difficult.

  \paragraph*{Methods:} The software outlined "Scaffolder" is implemented in
  the Ruby programming language and can be installed via the RubyGems software
  management system. Genome scaffolds are defined using YAML - a markup
  language, which is both human and machine-readable. Command line binaries and
  extensive documentation are provided.

  \paragraph*{Results:} This software allows a genome build to be defined in
  terms of the constituent contigs using a relatively simple to write and edit
  scaffold file syntax. The Scaffolder syntax further allows inserts to be
  added to the scaffold for the purpose of filling unknown sequence regions.
  Defining the genome construction in a file makes the scaffolding process
  reproducible and comparatively easier to edit than raw fasta.

  \paragraph*{Conclusions:} Scaffolder is easy to use genome scaffolding
  software. This tool promotes reproducibility and maintenance in building
  a genome. Scaffolder can be found at \scaffolder.

\end{abstract}

\ifthenelse{\boolean{publ}}{\begin{multicols}{2}}{}

% }}}
% {{{ Background
\section*{Background}
 Text for this section.\cite{koon,oreg,khar,zvai,xjon,schn,pond,smith,marg,hunn,advi,koha,mouse}
% }}}
% {{{ Implementation
%%%%%%%%%%%%%%%%%%%%%%
\section*{Implementation}
  Text for this section \ldots
% }}}
\section*{Results and Discussion} %{{{1

\subsection*{Scaffolder Simplifies Genome Finishing} %{{{2

The Scaffolder software allows manual or computational joining of multiple
discontiguous nucleotide sequences together into a contiguous super sequence.
Simple plain-text configuration files are used to select the nucleotide
sequences in the scaffold and how they should be joined together. This is
called the ``scaffold file''. \pb

The scaffold file is read by the scaffolder software to determine how the
scaffold super sequence is generated. In addition to specifying which
nucleotides should be used to generate the draft super sequence, the scaffold
file allows the sequences to trimmed down to smaller sub-sequences and to
reverse complement the sequence if necessary. \pb

There may be unknown regions in the sequences used in the scaffold. Region of
nucleotide sequence the assembler was unable to determine the nucleotide
sequence. The process of genome finishing may involve determining the correct
nucleotide sequence for these regions then filling the gaps with the additional
correct sequence. The scaffold file specifically allows these inserts to be
added to scaffold. These inserts can be treated the same way as a larger
contig: trimmed and/or reversed to match the required gap region. \pb

When scaffolding a genome there may be regions of sequence approximate size
region may be estimated from mate-pair insert size or from mapping the contigs
to reference genome. These unknown regions are useful to join non-contiguous
sequences together by the estimated size of the region whilst still
highlighting areas in the genome to be resolved. These unknown regions can be
specified in the scaffold file for the above reasons. \pb

The original raw nucleotide sequence for each assembled contig or insert
sequence is maintained in a fasta file. The fasta file of contigs is a common
output for assemblers and as such should be readily available after assembly
without additional processing. The choice of the nucleotide sequence to use in
the scaffold is specified using the first space-delimited word from the fasta
header of the sequence. By maintaining the nucleotide sequences in a separate
file this preserves the original raw sequence in the fasta file. Furthermore
this separates the actual nucleotide sequence from the determination of how it
is selected and edited in the final genome sequence. \pb

\subsection*{Defining a scaffold in a text file} %{{{2

The scaffold file is written using `YAML', a data format designed to be both
machine readable and manually editable. An example of a scaffold file in YAML
format is shown in Figure 1. This example scaffold file illustrates the
available features in scaffolder. The basic layout of the scaffold file is as
a list - a series of entries where each entry defines a region of sequence to
be generated for the final super sequence. In the final super sequence each of
the entries are joined together to generate a single continuous sequences. \pb

\subsubsection*{Simple sequence region} %{{{3

The first line of the scaffold file begins with three dashes. This is
a requirement to indicate the start of a YAML formatted document. Line 2 begins
with a dash character `-' denoting the first entry in the sequence scaffold.
Line 3 is indented by two spaces. Whitespace is used to delimite attributes in
the file. Lines with an incremented level of white space are considered linked
to the first preceeding non-indented line. \pb

The ``sequence'' tag used on line 3 indicates that this region is in the
scaffold will contain nucleotide sequence from the fasta file. The following
line indented by two spaces and indicating an attribute of the sequence entry
defined on line 3. In this case the ``source'' keyword identifies the name of
the sequence to be inserted into the scaffold. As described above the first
space-delimited word of the fasta header should be used to identify the
sequence. Taken together lines 2-4 describe the first entry in the scaffold is
a simple sequence which the identifier `sequence'. \pb

\subsubsection*{Unresolved sequence region} %{{{3

Line 5 has no whitespace indentation and begins with a dash and therefore
describes the next entry in the scaffold. This entry is identified by the
``unresolved'' tag, and specifies a region of unknown nucleotides but known
length. The following line specifies the size of this unknown region. Lines 5-6
therefore insert a region of 20 `N' characters in the super sequence directly
following the region defined by the first entry on lines 2-5. \pb

\subsubsection*{Trimmed sequence region with multiple inserts} %{{{3

The last entry adds the sequence 'sequence2' in the scaffold file. This entry
demonstrates three further attributes for a sequence entry.The `start' and
`stop' tags are used to trim the sequence used in the scaffold to these
coordinates inclusively. The ``reverse'' tag is used to reverse complement this
region in the final super sequence. \pb

This region further includes the ``inserts'' tag to update the inserted
sequence with additional regions of sequence. These inserts are also added as
a YAML list with each insert listed with a dash on one line followed by the
remainder of the insert errata. \pb

Lines 16-19 illustrate an insert using similar attributes to that of a sequence
entry. The reverse, start and stop tags trim the inserted sequence and the
`source' tag identifies the corresponding fasta sequence. The ``open'' and
``close'' tags on lines 20-21 determine where the insert is added to the
enclosing sequence.  These coordinates inclusively replace the region within
these coordinates with the specified fasta sequence. The open and close
positions need not be the same size as the inserted sequence. \pb

The next insert on lines 23-24 specifies only the open position. This
illustrate that only the one of the `open' or `close' tags is required.  When
only open or close tag is used then the opposing position is calculated from
the length on the insert fasta sequence. This allows inserts to partially fill
gap regions in sequences. \pb

\subsubsection*{Scaffolder file attribute processing} %{{{3

Each scaffolder sequence entry can have multiple attributes which applied in
different order can result in a different sequence. For instance reverse
complementing the sequence then adding inserts will have a different effect
compared with performing these actions in the reverse order. Scaffolder
therefore follows the following protocol for processing each entry in the
scaffold file. \pb

\paragraph{Reverse sort inserts:} If the entry is a sequence and has sequence
inserts these are reverse sorted by insert close position. The insert with the
largest close position is first and and the smallest close position is last.
This is done to preserve the the insert open/close coordinates as they are
added to the scaffold. \pb

\paragraph{Update sequence with insert:} Each insert is added to the sequence
replacing the region specified by the open an close position with the specified
source sequence.\pb

\paragraph{Update sequence start/stop co-ordinates:} If the insert added to the
sequence is a different length than the region replaced the start and stop
co-ordinates of the sequence are correspondingly updated if necessary. E.g. if
a 7 bp insert filled a bp region this would result in a 2bp difference in the
length of the sequence. The stop coordinated would be updated to reflect this.
\pb

\paragraph{Sub-sequence selected:} If the sequence is has start or stop
sequences specified these are used to trip the sequence to these co-ordinates.
\pb 

\paragraph{Sub-sequence reverse complemented: } If specified with the reverse keyword the sequence is reverse complemented.
\pb

\subsection*{Scaffolder usage} %{{{2

Scaffolder provides a ruby API that allows the scaffold and designated
sequences to be manipulated through a ruby programme. This allows scaffolder to
be integrated into exiting genomics works flows such as using Rake.  Scaffolder
all provides a command line binary for the manipulation of the scaffold. This
binary behaves as a standard Unix command line interface returning appropriate
exit statuses and providing manual pages for each of the commands. These two
methods are outlined in the following three sections. \pb

\subsubsection*{Installation} %{{{3

Scaffolder relies on both the ruby programming language and the RubyGems
package management system. These are commonly installed on recent versions of
Mac OSX and Linux distributions. The installation of these can be tested on the
command line by typing the following commands. \pb

\begin{verbatim}
ruby -v
gem -v
\end{verbatim}

If both these commands return without error this indicates the necessary
requirements to install scaffolder. The scaffolder API and command line tools
can then be installed with the follow command.

\begin{verbatim}
gem install scaffolder scaffolder-tools
\end{verbatim}

Finally to ensure that the command line tools are accessible these the
RubyGems binary directory should be in the command line path. This can be done
by adding the following to the ~/.bashrc file or other corresponding shell
configure file. Assuming that ~/.gem is the install path for RubyGems.

\begin{verbatim}
export PATH=~/.gem/bin:$PATH
\end{verbatim}

\subsubsection*{Command line interface} %{{{3

Scaffolder provides a command line interface to validate and build a genome
sequence from a scaffold. Assuming scaffolder has been installed and the
binary is available in the command line path the following command can be used
to invoke scaffolder. This will list the available commands for running
scaffolder. \pb

\begin{verbatim}
$ scaffolder
\end{verbatim}

Help on each command can be found by type `help' followed by the name of the
command. This will return the Unix manual page for the command of interest.
For example the following command will return the manual page for sequence
listing the available options for running the sequence command. \pb

\begin{verbatim}
$ scaffolder help sequence
\end{verbatim} 

To the generate the fasta sequence from a scaffold the sequence command should
be used. This should be passed the name of the file with containing the YAML
formatted scaffold and the corresponding fasta file of sequences. This will
then return the complete drafted sequence to the standard out. \pb

\begin{verbatim}
$ scaffolder sequence scaffold.yml sequences.fna
\end{verbatim}

Scaffolder also provides a function to test the scaffold for overlapping
insert sequence co-ordinates. As described in the section above overlapping
inserts will result in unexpected behaviour as the coordinates for the first
insert will change the expected insert location for the second insert. The
validate command will check the scaffold for any overlapping inserts and
return them as a YAML formatted list describing each one. \pb

\begin{verbatim}
$ scaffolder validate scaffold.yml sequences.fna
\end{verbatim}

\subsubsection*{Application programming interface} %{{{3

The scaffolder tool has a defined application programming interface (API) to
allow ruby programs to manipulate the scaffold. Detailing the complete API
would be overwhelming to this document. The scaffolder API is fully documented
online using at the scaffolder website. The following code snippet gives and
outline of how the API may be used to create a script to replicate the
function of the `scaffolder sequence' command described above. \pb

\begin{verbatim}
  #!/usr/bin/env ruby

  require 'rubygems'
  require 'scaffolder'

  scaffold_file = ARGV.shift
  sequence_file = ARGV.shift

  scaffold = Scaffolder.new(
    YAML.load(File.read(scaffold_file)),
    sequence_file)

  sequence = scaffold.map{|entry| entry.sequence }.join

  puts sequence
\end{verbatim}

This program illustrates that the scaffold object is initialised with two arguments: the YAML loaded scaffold file and the location of the fasta formatted sequence file. The scaffold is then enumerable as the next line illustrates the trivial code required to generate the scaffold. Further details of the attributes for each entry in the scaffold can be found in the documentation for the Scaffolder::Region class. \pb

\subsection*{Discussion of scaffolder approach} %{{{2

The main aims of scaffolder are two fold. The first is usability and the second is reproducibility.

\subsubsection{Simple to use}

Scaffolder is designed to be as simple to use as possible. Assuming the
existence of Ruby and RubyGems, scaffolder can be installed with a single
command. Scaffolder uses a minimal and compact syntax to described the
scaffold. This syntax is simple to construct and edit whilst being succinct and
readable in describing how the scaffold was constructed. Scaffolder provides
extensive documentation both for the command line tools and for the Scaffolder
Ruby API.

\subsubsection{Reproducible genome construction}

Constructing a genome sequence using a plain text scaffold file means the same
sequence is always output for the same file. The sequence produced from
manually joining separate contigs in a text editor can not be reliably
reproduced. Constructing the scaffold in a plain text file means that the
sequence is reproducible. Furthermore the source of each region in final
sequence can be identified. The scaffold file provides a readable record of
how the sequence was constructed. The scaffold file also means the scaffold is
easier to edit once already constructed. Much more difficult to do when working
on the sequence alone. The YAML text format is also much easier to compare
differences using standard diff commands. This makes scaffold files much more
simple to store in version control and to see the stages in the construction
of the genome sequence.

\section*{Conclusions} %{{{1

Scaffolder follows the Unix philosophy ``Do one thing and do it well''.
Scaffolder focuses on creating a simple to use and install tool for generating
a genome scaffold. The YAML data format for creating a scaffold file is
standardised and easily manipulated programmatically. This thereby follows
a second Unix tenet: ``If your data structures are good enough, the algorithm
to manipulate them should be trivial.''

\section*{Availability and Requirements} %{{{1

  \begin{description}
    \item[Project name:] Scaffolder
    \item[Project home page:] \scaffolder
    \item[Operating system:] Platform Independent. Tested on Mac OS X and
    Debian.
    \item[Programming language:] Ruby
    \item[Other requirements:] RubyGems
    \item[License:] MIT
    \item[Any restrictions to use by non-academics:] None
  \end{description}

% }}}
% {{{ Competing Interests
\section*{Authors contributions}
  The authors declare no competing interests.
% }}}
% {{{ Author Contributions
\section*{Authors contributions}

MDB developed and maintains the scaffolder tool. MDB and HAB wrote the
manuscript.

% }}}
% {{{ Author's Information
\section*{Authors contributions}
    Text for this section \ldots
% }}}
% {{{ Acknowledgements
\section*{Acknowledgements}
  \ifthenelse{\boolean{publ}}{\small}{}
  Text for this section \ldots
%{{{1 Bibliography
{\ifthenelse{\boolean{publ}}{\footnotesize}{\small}
 \bibliographystyle{bmc_article}  % Style BST file
  \bibliography{article} }     % Bibliography file (usually '*.bib' ) 

\ifthenelse{\boolean{publ}}{\end{multicols}}{}

\section*{Figures} %{{{1

  \subsection*{Figure 1 - Sample figure title}

  A short description of the figure content should go here.

\section*{Tables} %{{{1

  \subsection*{Table 1 - Sample table title}
    Here is an example of a \emph{small} table in \LaTeX\ using  
    \verb|\tabular{...}|. This is where the description of the table 
    should go. \par \mbox{}
    \par
    \mbox{
      \begin{tabular}{|c|c|c|}
        \hline \multicolumn{3}{|c|}{My Table}\\ \hline
        A1 & B2  & C3 \\ \hline
        A2 & ... & .. \\ \hline
        A3 & ..  & .  \\ \hline
      \end{tabular}
      }
  \subsection*{Table 2 - Sample table title}
    Large tables are attached as separate files but should
    still be described here.

\section*{Additional Files} %{{{1

  \subsection*{Additional file 1 --- Sample additional file title}
    Additional file descriptions text (including details of how to
    view the file, if it is in a non-standard format or the file extension).  This might
    refer to a multi-page table or a figure.

  \subsection*{Additional file 2 --- Sample additional file title}
    Additional file descriptions text.


\end{bmcformat}

\end{document}
