% {{{ Preamble


%% BioMed_Central_Tex_Template_v1.05
%%                                      %
%  bmc_article.tex            ver: 1.05 %
%                                       %


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                     %%
%%  LaTeX template for BioMed Central  %%
%%     journal article submissions     %%
%%                                     %%
%%         <27 January 2006>           %%
%%                                     %%
%%                                     %%
%% Uses:                               %%
%% cite.sty, url.sty, bmc_article.cls  %%
%% ifthen.sty. multicol.sty		       %%
%%									   %%
%%                                     %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                                                 %%	
%% For instructions on how to fill out this Tex template           %%
%% document please refer to Readme.pdf and the instructions for    %%
%% authors page on the biomed central website                      %%
%% http://www.biomedcentral.com/info/authors/                      %%
%%                                                                 %%
%% Please do not use \input{...} to include other tex files.       %%
%% Submit your LaTeX manuscript as one .tex document.              %%
%%                                                                 %%
%% All additional figures and files should be attached             %%
%% separately and not embedded in the \TeX\ document itself.       %%
%%                                                                 %%
%% BioMed Central currently use the MikTex distribution of         %%
%% TeX for Windows) of TeX and LaTeX.  This is available from      %%
%% http://www.miktex.org                                           %%
%%                                                                 %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


\NeedsTeXFormat{LaTeX2e}[1995/12/01]
\documentclass[10pt]{bmc_article}    



% Load packages
\usepackage{cite} % Make references as [1-4], not [1,2,3,4]
\usepackage{url}  % Formatting web addresses  
\usepackage{ifthen}  % Conditional 
\usepackage{multicol}   %Columns
\usepackage[utf8]{inputenc} %unicode support
%\usepackage[applemac]{inputenc} %applemac support if unicode package fails
%\usepackage[latin1]{inputenc} %UNIX support if unicode package fails
\urlstyle{rm}
 
 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%	
%%                                             %%
%%  If you wish to display your graphics for   %%
%%  your own use using includegraphic or       %%
%%  includegraphics, then comment out the      %%
%%  following two lines of code.               %%   
%%  NB: These line *must* be included when     %%
%%  submitting to BMC.                         %% 
%%  All figure files must be submitted as      %%
%%  separate graphics through the BMC          %%
%%  submission process, not included in the    %% 
%%  submitted article.                         %% 
%%                                             %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%                     

\def\includegraphic{}
\def\includegraphics{}

\setlength{\topmargin}{0.0cm}
\setlength{\textheight}{21.5cm}
\setlength{\oddsidemargin}{0cm} 
\setlength{\textwidth}{16.5cm}
\setlength{\columnsep}{0.6cm}

\newboolean{publ}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                              %%
%% You may change the following style settings  %%
%% Should you wish to format your article       %%
%% in a publication style for printing out and  %%
%% sharing with colleagues, but ensure that     %%
%% before submitting to BMC that the style is   %%
%% returned to the Review style setting.        %%
%%                                              %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%Review style settings
\newenvironment{bmcformat}{\begin{raggedright}\baselineskip20pt\sloppy\setboolean{publ}{false}}{\end{raggedright}\baselineskip20pt\sloppy}

%Publication style settings
%\newenvironment{bmcformat}{\fussy\setboolean{publ}{true}}{\fussy}

% Scaffolder url
\urldef{\scaffolder}\url{http://www.michaelbarton.me.uk/projects/scaffolder/}

% Begin ...
\begin{document}
\begin{bmcformat}
% }}}
% {{{ Title

\title{Scaffolder - Software for Microbial Genome Scaffolding.}

\author{
  Michael D Barton$^{1}$%
  \email{Michael D Barton - mail@michaelbarton.me.uk}%
\and
  Hazel A Barton\correspondingauthor$^1$%
  \email{Hazel A Barton\correspondingauthor - bartonh@nku.edu}%
      }

\address{\iid(1) Department of Biological Sciences, Northern Kentucky%
University, Nunn Drive, Highland Heights, KY 41076 }%

\maketitle
%}}}
% {{{ Abstract

\begin{abstract}

  \paragraph*{Background:} Assembly of short read sequencing data can result in
  a fragmented series of contigs. Therefore a common step in a genome project
  is to join neighbouring sequence regions together and fill gaps using insert
  data generated through PCR. This scaffolding step, however, is non-trivial
  and requires manually editing large blocks of nucleotide sequence. Joining
  sequence contigs together also hides the source of each region in the final
  genome sequence. Taken together, these considerations may make reproducing or
  editing an existing genome build difficult.

  \paragraph*{Methods:} The software outlined ``Scaffolder'' is implemented in
  the Ruby programming language and can be installed via the RubyGems software
  management system. Genome scaffolds are defined using YAML - a markup
  language, which is both human and machine-readable. Command line binaries and
  extensive documentation are provided.

  \paragraph*{Results:} This software allows a genome build to be defined in
  terms of the constituent contigs using a relatively simple to write and edit
  scaffold file syntax. The Scaffolder syntax further allows inserts to be
  added to the scaffold for the purpose of filling unknown sequence regions.
  Defining the genome construction in a file makes the scaffolding process
  reproducible and comparatively easier to edit than raw fasta.

  \paragraph*{Conclusions:} Scaffolder is easy to use genome scaffolding
  software. This tool promotes reproducibility and maintenance in building
  a genome. Scaffolder can be found at \scaffolder.

\end{abstract}

\ifthenelse{\boolean{publ}}{\begin{multicols}{2}}{}

% }}}
<<<<<<< HEAD
\section*{Background} %{{{1
%{{{2

High throughput sequencing can produce hundreds of thousands to millions of
sequence reads from a genome. Each read represents the machine-interpreted
nucleotide sequence from a short region in the genome. At the time of writing
high throughput sequencing is limited to producing reads less than $>$1000
nucleotides. Therefore to produce to produce a sequence longer than this
length, e.g. a complete genome, these numerous individual reads must be pieced
together. The process of joining reads into longer sequences is the `assembly'
stage in a genome project \cite{assembly-reviews}. \pb

Assembly software takes the nucleotide reads produced by the sequencing
hardware and, in the ideal case, outputs a complete genome sequence compose of
these individual reads. An analogy to describe this is a jigsaw puzzle. Each
nucleotide read represents a single piece and the final genome sequence is the
completed puzzle. Repetitive nucleotide `repeat' regions within a genome or
biased and incomplete sequencing data may however prevent the genome being
assembled into a continuous sequence. This is analogous to missing jigsaw
pieces or pieces that fit to multiple other pieces. \pb

The recent advent of these current high throughput sequencing methods has lead
to a renewed interest into algorithms for solving the problem of genome
assembly \cite{assembly-reviews}. Genome assembly software is often unable to
completely these sequence reads together due to problems outlined above.
Instead assembly algorithms will generate large contiguous regions of sequence
(`contigs') composed of many individual reads. These represent optimum assembly
of reads into the longest sequences for the assembly algorithm. \pb

The process of ameliorating a draft sequence to a finished continuous sequence
can be expensive in time and laboratory effort. Therefore in many cases a set
of contigs may often describe enough of the genomic content for many research
questions \cite{branscomb2002}. A continuous high-quality `finished' genome
sequence does however provide further depth of information
\cite{parkhill2002,fraser2002}. The benefits of a complete genome sequence
include complete resolution of repeat regions and the exact distances between
genomic elements. This process of joining contigs from assembly together to form
a continuous sequence may be called `scaffolding' or `finishing'. Here on we
use the term scaffolding. \pb

\subsection*{Scaffolding} %{{{2

Scaffolding is the process of joining a series disconnected contigs into
complete continuous sequence. Due to genomic complexity, missing data and
diminishing returns scaffolding may not ultimately produce a final completed
sequence. The process of scaffolding may still however joining a subset of
contigs or additional regions of sequence to fill gaps.

\subsubsection*{Contig Orientation} %{{{3

The sequencing process produces reads from either strand of the DNA helix. The
resulting contigs constructed from these reads therefore point in either of two
DNA strand directions. Contig orientation involves reverse complementing
sequences so that all contigs point in the same direction and represent the
same DNA strand. In the case of microbes this orientation is often the 5`-3`
strand leading away from the origin of replication.

\subsubsection*{Contig Ordering} %{{{3

Contig ordering determines the placement of contigs to best represent the
actual nucleotides as appearing in the genome sequence. This may also determine
if any contigs represent extra-genomic DNA, such as plasmids, and therefore
should be treated separately from genomic contigs. The contig order is commonly
referenced to the origin of replication where the contig containing the
replication start point is first. All subsequent contigs then follow in the
direction of DNA replication.

\subsubsection*{Contig Distancing} %{{{3

Given the order and orientation of contigs determining the distances between
contigs may be useful for estimating genome size. The size of each gaps between
contigs represents the length of an unknown region in the genome. Filling in
these regions in with unknown nucleotide characters (`N') allows a draft
continuous representation of the genome to produced. This draft scaffold
sequence represents both the contigs and unknown regions of sequence between
them.

\subsubsection*{Gap Closing and Filling} %{{{3

During the scaffolding process closing and filling sequence gaps improves the
accuracy of the genome scaffold. This may require returning to the laboratory
to perform additional sequencing or using computational methods. Once all
contigs have been joined and gaps in a scaffold closed the genome may be
considered as finished. In the case of organisms or particular interest
additional sequencing may be performed to further `proof' any low coverage
regions or detect erroneous nucleotides resulting from insertions, deletions or
substitutions.

\subsection*{Computational Methods for Scaffolding} %{{{2

The process of resolving a genome sequence and building a scaffold uses
\emph{in vitro} methods, \emph{in silico} methods, or both in conjunction.
Examples of computation methods is the use of pair reads data, where sequencing
reads are a known distance apart in the genome, or from comparing contigs to
a reference genome. Laboratory methods may use PCR to amplify the DNA in gap
regions, then use traditional Sanger sequencing to determine this sequence.
Computation methods are more preferable as they less costly in laboratory time
and materials compared to manual gap resolution \cite{nagarajan2010}. \pb

The need for computational scaffolding methods has resulted in a variety of
algorithmic approaches. Synteny methods compare the assembled contigs to a more
complete reference genome sequence. Areas of corresponding sequence between the
the two sets are indicative of the possible contig organisation
\cite{richter2007,zhao2008}. \pb

Repeat regions may be responsible for many of the gaps in building a genome
sequence. The existence of tandemly repeated nucleotide regions produces many
reads of similar sequence. The assembly of these regions may result in the
repeats being collapsed into a single instance or ignored by conservative
assembly algorithms. These regions may be reassembled using algorithms
specifically for repeat resolution \cite{mulyukov2002,koren2010}. A related
approach uses uniquely matching reads try find overlaps and bridge across the
gaps between contigs \cite{tsai2010}. \pb

As mentioned above, the genome may be prepared prior to sequencing so that
reads are produced in pairs at a known distance apart. This paired-read data
provides an extra level of information on how contigs may be scaffolded
together. Heuristic graph algorithms may be used to search for optimal
configuration of contigs in the scaffold to take into account this information
\cite{dayarian2010}. Additional data using a reference genome may also be
combined when estimating the best contig configuration \cite{pop2004}. \pb

Finally when the scaffold cannot be completely resolved \emph{in silico}
software packages exist to suggest the likely primers necessary for determining
the sequence in gap regions \cite{gordon2001,nagarajan2010}. \pb

\subsection*{Summary} %{{{2

Taken together these methods provide a wide variety of possible approaches for
scaffolding a genome. Scaffolding however may still however require manual
editing of the scaffold if computational methods do not resolve the final
sequence.  This process involves moving and joining larges blocks of nucleotide
text together. This however may be error-prone and manual editing also
precludes reproducibility of the genome scaffold. \pb

The software outlined here ``Scaffolder'' is aimed at creating a syntax and
software framework for editing a genome scaffold. Scaffolder uses a specific
data syntax to define how contigs are joined, additional sequences inserted,
and specify unknown regions. This syntax allows a scaffold to be edited in much
easier way and makes genome scaffolds reproducible. \pb

\section*{Implementation} %{{{1

\subsection*{Code and Dependencies} %{{{2

Scaffolder is written in the Ruby programming language, version 1.8.7
\cite{ruby-lang}. Unit tests were implemented during development to maintain
functionality of individual elements of source code. Integration tests were
also written to test the scaffold software interface as a whole. These unit and
integration tests were written using the RSpec and Cucumber libraries
respectively\cite{rspec}.\pb

The Scaffolder source code is documented using the Yard library\cite{yard}.
Unix manual pages for the command line interface were generated using the Ronn
library\cite{ronn}. The manipulation of biological sequences in Scaffolder is
performed using the BioRuby library\cite{bioruby}. A full list of the software
dependencies in scaffolder can be found in the Gemfile in the root of source
code directory.\pb

\subsection*{Scaffold file syntax} %{{{2

The choice of nucleotide sequences comprising the genome scaffold is specified
in YAML \cite{yaml}. YAML is a language format for data using white-space and
indentation to produce a machine readable structure. As a standardised data
format third-party developers have the option to generate a genome scaffold
using any programming language for which a YAML parser exists. Furthermore YAML
formatted scaffold files can be validated using third party tools such as
Kwalify\cite{kwalify}.\pb

The initial assembly of sequencing reads may result in an incomplete genome
build. Comparison of contigs to reference genomes and the generation of
additional sequence from PCR translates to genome finishing as an ongoing
process. Therefore the scaffold file should be simple to edit manually in
addition to being computationally tractable. This requirement for the scaffold
files to be modifiable in a simple simple text editor is the second reason YAML
was selected as the format for scaffold files.\pb

The form of the scaffold file is a list of entries. Each entry corresponds to
a length of sequence in the the draft genome sequence. When generating the
genome sequence from the scaffold file each entry is processed in order. Each
of the nucleotide sequence specified by the scaffold file entries are
concatenated together to produce the draft genome sequence. Each entry in the
scaffold file may have attributes that define whether a sub-sequence or the
reverse complement of the sequence should be used. The types of attributes are
outlined in the Results and Discussion section below.\pb

% {{{ Results and Discussion
 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% Results and Discussion %%
%%
\section*{Results and Discussion}
  \subsection*{Results sub-heading}
    \subsubsection*{This is a sub-sub-heading}
      Sub-sub-sub-headings are made with the \textsl{\\subsubsection} command. \pb
      pb at end of lines ensures correct paragraph spacing.\pb
	  Text for this sub-sub-section \ldots
    \subsubsection*{Another sub-sub-sub-heading}
      Text for this sub-sub-section \ldots

% {{{ Conclusions
%%%%%%%%%%%%%%%%%%%%%%
\section*{Conclusions}
  Text for this section \ldots
% }}}
% {{{ Availability and Requirements
\section*{Availability and Requirements}

  \begin{description}
    \item[Project name:] Scaffolder
    \item[Project home page:] \scaffolder
    \item[Operating system:] Platform Independent. Tested on Mac OS X and
    Debian.
    \item[Programming language:] Ruby
    \item[Other requirements:] RubyGems
    \item[License:] MIT
    \item[Any restrictions to use by non-academics:] None
  \end{description}

% }}}
% {{{ Competing Interests
\section*{Authors contributions}
  The authors declare no competing interests.
% }}}
% {{{ Availability and Requirements
\section*{List of Abbreviations Used}

  \begin{description}
    \item[YAML:] YAML Ain't Markup Language\cite{yaml}
  \end{description}

% }}}
% {{{ Author Contributions
\section*{Authors contributions}

MDB developed and maintains the scaffolder tool. MDB and HAB wrote the
manuscript.

% }}}
% {{{ Author's Information
\section*{Authors contributions}
    Text for this section \ldots
% }}}
% {{{ Acknowledgements
\section*{Acknowledgements}
  \ifthenelse{\boolean{publ}}{\small}{}
  Text for this section \ldots
%}}}
% {{{ Bibliography
{\ifthenelse{\boolean{publ}}{\footnotesize}{\small}
 \bibliographystyle{bmc_article}  % Style BST file
  \bibliography{article} }     % Bibliography file (usually '*.bib' ) 
%}}}
% {{{ Figures
\ifthenelse{\boolean{publ}}{\end{multicols}}{}

\section*{Figures}
  \subsection*{Figure 1 - Sample figure title}
      A short description of the figure content
      should go here.

  \subsection*{Figure 2 - Sample figure title}
      Figure legend text.
% }}}
% {{{ Tables
\section*{Tables}
  \subsection*{Table 1 - Sample table title}
    Here is an example of a \emph{small} table in \LaTeX\ using  
    \verb|\tabular{...}|. This is where the description of the table 
    should go. \par \mbox{}
    \par
    \mbox{
      \begin{tabular}{|c|c|c|}
        \hline \multicolumn{3}{|c|}{My Table}\\ \hline
        A1 & B2  & C3 \\ \hline
        A2 & ... & .. \\ \hline
        A3 & ..  & .  \\ \hline
      \end{tabular}
      }
  \subsection*{Table 2 - Sample table title}
    Large tables are attached as separate files but should
    still be described here.
% }}}
% {{{ Additional Files
\section*{Additional Files}
  \subsection*{Additional file 1 --- Sample additional file title}
    Additional file descriptions text (including details of how to
    view the file, if it is in a non-standard format or the file extension).  This might
    refer to a multi-page table or a figure.

  \subsection*{Additional file 2 --- Sample additional file title}
    Additional file descriptions text.


\end{bmcformat}
\end{document}
% }}}
