% {{{ Preamble


%% BioMed_Central_Tex_Template_v1.05
%%                                      %
%  bmc_article.tex            ver: 1.05 %
%                                       %


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                     %%
%%  LaTeX template for BioMed Central  %%
%%     journal article submissions     %%
%%                                     %%
%%         <27 January 2006>           %%
%%                                     %%
%%                                     %%
%% Uses:                               %%
%% cite.sty, url.sty, bmc_article.cls  %%
%% ifthen.sty. multicol.sty		       %%
%%									   %%
%%                                     %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                                                 %%	
%% For instructions on how to fill out this Tex template           %%
%% document please refer to Readme.pdf and the instructions for    %%
%% authors page on the biomed central website                      %%
%% http://www.biomedcentral.com/info/authors/                      %%
%%                                                                 %%
%% Please do not use \input{...} to include other tex files.       %%
%% Submit your LaTeX manuscript as one .tex document.              %%
%%                                                                 %%
%% All additional figures and files should be attached             %%
%% separately and not embedded in the \TeX\ document itself.       %%
%%                                                                 %%
%% BioMed Central currently use the MikTex distribution of         %%
%% TeX for Windows) of TeX and LaTeX.  This is available from      %%
%% http://www.miktex.org                                           %%
%%                                                                 %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


\NeedsTeXFormat{LaTeX2e}[1995/12/01]
\documentclass[10pt]{bmc_article}    



% Load packages
\usepackage{cite} % Make references as [1-4], not [1,2,3,4]
\usepackage{url}  % Formatting web addresses  
\usepackage{ifthen}  % Conditional 
\usepackage{multicol}   %Columns
\usepackage[utf8]{inputenc} %unicode support
%\usepackage[applemac]{inputenc} %applemac support if unicode package fails
%\usepackage[latin1]{inputenc} %UNIX support if unicode package fails
\urlstyle{rm}
 
 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%	
%%                                             %%
%%  If you wish to display your graphics for   %%
%%  your own use using includegraphic or       %%
%%  includegraphics, then comment out the      %%
%%  following two lines of code.               %%   
%%  NB: These line *must* be included when     %%
%%  submitting to BMC.                         %% 
%%  All figure files must be submitted as      %%
%%  separate graphics through the BMC          %%
%%  submission process, not included in the    %% 
%%  submitted article.                         %% 
%%                                             %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%                     

\def\includegraphic{}
\def\includegraphics{}

\setlength{\topmargin}{0.0cm}
\setlength{\textheight}{21.5cm}
\setlength{\oddsidemargin}{0cm} 
\setlength{\textwidth}{16.5cm}
\setlength{\columnsep}{0.6cm}

\newboolean{publ}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                              %%
%% You may change the following style settings  %%
%% Should you wish to format your article       %%
%% in a publication style for printing out and  %%
%% sharing with colleagues, but ensure that     %%
%% before submitting to BMC that the style is   %%
%% returned to the Review style setting.        %%
%%                                              %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%Review style settings
\newenvironment{bmcformat}{\begin{raggedright}\baselineskip20pt\sloppy\setboolean{publ}{false}}{\end{raggedright}\baselineskip20pt\sloppy}

%Publication style settings
%\newenvironment{bmcformat}{\fussy\setboolean{publ}{true}}{\fussy}

% Scaffolder url
\urldef{\scaffolder}\url{http://www.michaelbarton.me.uk/projects/scaffolder/}

% Begin ...
\begin{document}
\begin{bmcformat}
% }}}
% {{{ Title

\title{Scaffolder - Software for Microbial Genome Scaffolding.}

\author{
  Michael D Barton$^{1}$%
  \email{Michael D Barton - mail@michaelbarton.me.uk}%
\and
  Hazel A Barton\correspondingauthor$^1$%
  \email{Hazel A Barton\correspondingauthor - bartonh@nku.edu}%
      }

\address{\iid(1) Department of Biological Sciences, Northern Kentucky%
University, Nunn Drive, Highland Heights, KY 41076 }%

\maketitle
%}}}
% {{{ Abstract

\begin{abstract}

  \paragraph*{Background:} Assembly of short read sequencing data can result in
  a fragmented series of contigs. Therefore a common step in a genome project
  is to join neighbouring sequence regions together and fill gaps using insert
  data generated through PCR. This scaffolding step, however, is non-trivial
  and requires manually editing large blocks of nucleotide sequence. Joining
  sequence contigs together also hides the source of each region in the final
  genome sequence. Taken together, these considerations may make reproducing or
  editing an existing genome build difficult.

  \paragraph*{Methods:} The software outlined "Scaffolder" is implemented in
  the Ruby programming language and can be installed via the RubyGems software
  management system. Genome scaffolds are defined using YAML - a markup
  language, which is both human and machine-readable. Command line binaries and
  extensive documentation are provided.

  \paragraph*{Results:} This software allows a genome build to be defined in
  terms of the constituent contigs using a relatively simple to write and edit
  scaffold file syntax. The Scaffolder syntax further allows inserts to be
  added to the scaffold for the purpose of filling unknown sequence regions.
  Defining the genome construction in a file makes the scaffolding process
  reproducible and comparatively easier to edit than raw fasta.

  \paragraph*{Conclusions:} Scaffolder is easy to use genome scaffolding
  software. This tool promotes reproducibility and maintenance in building
  a genome. Scaffolder can be found at \scaffolder.

\end{abstract}

\ifthenelse{\boolean{publ}}{\begin{multicols}{2}}{}

% }}}
% {{{ Background
\section*{Background}
 Text for this section.\cite{koon,oreg,khar,zvai,xjon,schn,pond,smith,marg,hunn,advi,koha,mouse}
% }}}
% {{{ Implementation
%%%%%%%%%%%%%%%%%%%%%%
\section*{Implementation}
  Text for this section \ldots
% }}}
\section*{Results and Discussion} %{{{1

\subsection*{Scaffolder Description} %{{{2

Individual reads produced from nucleotide sequencing can be assembled into
contiguous regions (``contigs'') of many overlapping individual reads. These
reads may however not represent a complete genome sequence where all contigs do
not result in a single complete sequence. The software outlined here
``Scaffolder'' allows these contigs to be further joined together into
a continuous single sequence. This selection, order and direction of contigs
into a final draft genome is often described as the `scaffold'. Scaffolder
allows the manual or computational specification of a genome scaffold by
specifying which contig sequences are used in simple to edit configuration
files. 

The scaffold file specifies how contigs should be joined together in the draft
genome sequence. The raw nucleotide sequence for each contig is stored in
a fasta file, as might be produced from assembly of sequencing reads using
programs such as Newbler\cite{miller2010} or ABySS\cite{simpson2009}.
Scaffolder then the contig to be selected by specifying the name of the contig
in the scaffold file. The scaffold file also allows a sub-sequence or the
reverse complement of a contig to used. This therefore preserves the original
sequence data but also allowing sequences to manipulated in the final genome
build.

The majority of a genome can be estimated from scaffolding assembled contigs
into the most parsimonious build. There may however still be internal gaps in
the contigs from mate pair assembly based on estimated mate pair insert size.
Scaffolder allows inserts to be added to scaffolded so that gaps may be bridged
and filled - a common process in finishing a genome sequence. Inserts may be
treated the same way as a contig where the raw sequence, as may be produced by
PCR is stored in a fasta file, and then trimmed and/or reversed to fill the gap
in the contig.

Scaffolder allows the specification of unknown regions in the scaffold. These
are areas of the genome where the nucleotide sequence is unknown but the
approximate size region may be estimated from mate pair insert size or from
mapping contigs to reference genome and estimating the size of the gap. These
unknown regions allow a continuous draft sequence to be scaffolded but whilst
still highlighting regions waiting to be resolved.

\subsection*{Defining a scaffold in a text file} %{{{2

The genome scaffold is defined YAML format. YAML is an easy to edit data format
that is both machine readable and easy to edit manually.

INSERT FIGURE WITH EXAMPLE LAYOUT OF SCAFFOLD

  * Screen shot with example file
  * Illustrate three types of region in the scaffold file
  * Explanation of what each region does
  * Expected output from each region

Scaffolder provides a ruby API that allows the scaffold and designated
sequences to be manipulated through a ruby programme. This allows scaffolder to
be integrated into exiting genomics works flows such as using Rake. Scaffolder
all provides a command line binary for the manipulation of the scaffold. This
binary behaves as a standard Unix command line interface returning appropriate
exit statuses and providing manual pages for each of the commands. These two
methods are outlined in the following two sections.

\subsection*{Command line usage} %{{{2

  * Illustrate running command line to generate output
  * Demonstrate scaffold validate to test inserts overlap

\subsection*{Programmatic usage of API} %{{{2

  * Demonstrate how scaffolder can be manipulated programmatically
  * Examine scaffolder data using a ruby program
  * Example of extending sequence class with a new method \%GC content
  * Create a new class to do something

\subsection*{Installation} %{{{2

  * Install through RubyGems
  * Make sure the bin is in the class path

\section*{Conclusions} %{{{1
  Text for this section \ldots
% }}}
% {{{ Availability and Requirements
\section*{Availability and Requirements}

  \begin{description}
    \item[Project name:] Scaffolder
    \item[Project home page:] \scaffolder
    \item[Operating system:] Platform Independent. Tested on Mac OS X and
    Debian.
    \item[Programming language:] Ruby
    \item[Other requirements:] RubyGems
    \item[License:] MIT
    \item[Any restrictions to use by non-academics:] None
  \end{description}

% }}}
% {{{ Competing Interests
\section*{Authors contributions}
  The authors declare no competing interests.
% }}}
% {{{ Author Contributions
\section*{Authors contributions}

MDB developed and maintains the scaffolder tool. MDB and HAB wrote the
manuscript.

% }}}
% {{{ Author's Information
\section*{Authors contributions}
    Text for this section \ldots
% }}}
% {{{ Acknowledgements
\section*{Acknowledgements}
  \ifthenelse{\boolean{publ}}{\small}{}
  Text for this section \ldots
%}}}
% {{{ Bibliography
{\ifthenelse{\boolean{publ}}{\footnotesize}{\small}
 \bibliographystyle{bmc_article}  % Style BST file
  \bibliography{article} }     % Bibliography file (usually '*.bib' ) 
%}}}
% {{{ Figures
\ifthenelse{\boolean{publ}}{\end{multicols}}{}

\section*{Figures}
  \subsection*{Figure 1 - Sample figure title}
      A short description of the figure content
      should go here.

  \subsection*{Figure 2 - Sample figure title}
      Figure legend text.
% }}}
% {{{ Tables
\section*{Tables}
  \subsection*{Table 1 - Sample table title}
    Here is an example of a \emph{small} table in \LaTeX\ using  
    \verb|\tabular{...}|. This is where the description of the table 
    should go. \par \mbox{}
    \par
    \mbox{
      \begin{tabular}{|c|c|c|}
        \hline \multicolumn{3}{|c|}{My Table}\\ \hline
        A1 & B2  & C3 \\ \hline
        A2 & ... & .. \\ \hline
        A3 & ..  & .  \\ \hline
      \end{tabular}
      }
  \subsection*{Table 2 - Sample table title}
    Large tables are attached as separate files but should
    still be described here.
% }}}
% {{{ Additional Files
\section*{Additional Files}
  \subsection*{Additional file 1 --- Sample additional file title}
    Additional file descriptions text (including details of how to
    view the file, if it is in a non-standard format or the file extension).  This might
    refer to a multi-page table or a figure.

  \subsection*{Additional file 2 --- Sample additional file title}
    Additional file descriptions text.


\end{bmcformat}
\end{document}
% }}}
