% {{{ Preamble


%% BioMed_Central_Tex_Template_v1.05
%%                                      %
%  bmc_article.tex            ver: 1.05 %
%                                       %


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                     %%
%%  LaTeX template for BioMed Central  %%
%%     journal article submissions     %%
%%                                     %%
%%         <27 January 2006>           %%
%%                                     %%
%%                                     %%
%% Uses:                               %%
%% cite.sty, url.sty, bmc_article.cls  %%
%% ifthen.sty. multicol.sty		       %%
%%									   %%
%%                                     %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                                                 %%	
%% For instructions on how to fill out this Tex template           %%
%% document please refer to Readme.pdf and the instructions for    %%
%% authors page on the biomed central website                      %%
%% http://www.biomedcentral.com/info/authors/                      %%
%%                                                                 %%
%% Please do not use \input{...} to include other tex files.       %%
%% Submit your LaTeX manuscript as one .tex document.              %%
%%                                                                 %%
%% All additional figures and files should be attached             %%
%% separately and not embedded in the \TeX\ document itself.       %%
%%                                                                 %%
%% BioMed Central currently use the MikTex distribution of         %%
%% TeX for Windows) of TeX and LaTeX.  This is available from      %%
%% http://www.miktex.org                                           %%
%%                                                                 %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


\NeedsTeXFormat{LaTeX2e}[1995/12/01]
\documentclass[10pt]{bmc_article}    



% Load packages
\usepackage{cite} % Make references as [1-4], not [1,2,3,4]
\usepackage{url}  % Formatting web addresses  
\usepackage{ifthen}  % Conditional 
\usepackage{multicol}   %Columns
\usepackage[utf8]{inputenc} %unicode support
%\usepackage[applemac]{inputenc} %applemac support if unicode package fails
%\usepackage[latin1]{inputenc} %UNIX support if unicode package fails
\urlstyle{rm}
 
 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%	
%%                                             %%
%%  If you wish to display your graphics for   %%
%%  your own use using includegraphic or       %%
%%  includegraphics, then comment out the      %%
%%  following two lines of code.               %%   
%%  NB: These line *must* be included when     %%
%%  submitting to BMC.                         %% 
%%  All figure files must be submitted as      %%
%%  separate graphics through the BMC          %%
%%  submission process, not included in the    %% 
%%  submitted article.                         %% 
%%                                             %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%                     

\def\includegraphic{}
\def\includegraphics{}

\setlength{\topmargin}{0.0cm}
\setlength{\textheight}{21.5cm}
\setlength{\oddsidemargin}{0cm} 
\setlength{\textwidth}{16.5cm}
\setlength{\columnsep}{0.6cm}

\newboolean{publ}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                              %%
%% You may change the following style settings  %%
%% Should you wish to format your article       %%
%% in a publication style for printing out and  %%
%% sharing with colleagues, but ensure that     %%
%% before submitting to BMC that the style is   %%
%% returned to the Review style setting.        %%
%%                                              %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%Review style settings
\newenvironment{bmcformat}{\begin{raggedright}\baselineskip20pt\sloppy\setboolean{publ}{false}}{\end{raggedright}\baselineskip20pt\sloppy}

%Publication style settings
%\newenvironment{bmcformat}{\fussy\setboolean{publ}{true}}{\fussy}

% Scaffolder url
\urldef{\scaffolder}\url{http://www.michaelbarton.me.uk/projects/scaffolder/}

% Begin ...
\begin{document}
\begin{bmcformat}
% }}}
% {{{ Title

\title{Scaffolder - Software for Microbial Genome Scaffolding.}

\author{
  Michael D Barton$^{1}$%
  \email{Michael D Barton - mail@michaelbarton.me.uk}%
\and
  Hazel A Barton\correspondingauthor$^1$%
  \email{Hazel A Barton\correspondingauthor - bartonh@nku.edu}%
      }

\address{\iid(1) Department of Biological Sciences, Northern Kentucky%
University, Nunn Drive, Highland Heights, KY 41076 }%

\maketitle
%}}}
% {{{ Abstract

\begin{abstract}

  \paragraph*{Background:} Assembly of short read sequencing data can result in
  a fragmented series of contigs. Therefore a common step in a genome project
  is to join neighbouring sequence regions together and fill gaps using insert
  data generated through PCR. This scaffolding step, however, is non-trivial
  and requires manually editing large blocks of nucleotide sequence. Joining
  sequence contigs together also hides the source of each region in the final
  genome sequence. Taken together, these considerations may make reproducing or
  editing an existing genome build difficult.

  \paragraph*{Methods:} The software outlined "Scaffolder" is implemented in
  the Ruby programming language and can be installed via the RubyGems software
  management system. Genome scaffolds are defined using YAML - a markup
  language, which is both human and machine-readable. Command line binaries and
  extensive documentation are provided.

  \paragraph*{Results:} This software allows a genome build to be defined in
  terms of the constituent contigs using a relatively simple to write and edit
  scaffold file syntax. The Scaffolder syntax further allows inserts to be
  added to the scaffold for the purpose of filling unknown sequence regions.
  Defining the genome construction in a file makes the scaffolding process
  reproducible and comparatively easier to edit than raw fasta.

  \paragraph*{Conclusions:} Scaffolder is easy to use genome scaffolding
  software. This tool promotes reproducibility and maintenance in building
  a genome. Scaffolder can be found at \scaffolder.

\end{abstract}

\ifthenelse{\boolean{publ}}{\begin{multicols}{2}}{}

% }}}
\section*{Background} %{{{1
%{{{2

High throughput sequencing can produce hundreds of thousands to millions of
sequence reads from a genome. Each read represents the machine-interpreted
nucleotide sequence from a short region in the genome. At the time of writing
high throughput sequencing is limited to producing reads less than $>$1000
nucleotides. Therefore to produce to produce a sequence longer than this
length, e.g. a complete genome, these numerous individual reads must be pieced
together. The process of joining reads into longer sequences is the `assembly'
stage in a genome project \cite{assembly-reviews}. \pb

Assembly software takes the nucleotide reads produced by the sequencing
hardware and, in the ideal case, outputs a complete genome sequence compose of
these individual reads. An analogy to describe this is a jigsaw puzzle. Each
nucleotide read represents a single piece and the final genome sequence is the
completed puzzle. Repetitive nucleotide `repeat' regions within a genome or
biased and incomplete sequencing data may however prevent the genome being
assembled into a continuous sequence. This is analogous to missing jigsaw
pieces or pieces that fit to multiple other pieces. \pb

The recent advent of these current high throughput sequencing methods has lead
to a renewed interest into algorithms for solving the problem of genome
assembly \cite{assembly-reviews}. Genome assembly software is often unable to
completely these sequence reads together due to problems outlined above.
Instead assembly algorithms will generate large contiguous regions of sequence
(`contigs') composed of many individual reads. These represent optimum assembly
of reads into the longest sequences for the assembly algorithm. \pb

The process of ameliorating a draft sequence to a finished continuous sequence
can be expensive in time and laboratory effort. Therefore in many cases a set
of contigs may often describe enough of the genomic content for many research
questions \cite{branscomb2002}. A continuous high-quality `finished' genome
sequence does however provide further depth of information
\cite{parkhill2002,fraser2002}. The benefits of a complete genome sequence
include complete resolution of repeat regions and the exact distances between
genomic elements. This process of joining contigs from assembly together to form
a continuous sequence may be called `scaffolding' or `finishing'. Here on we
use the term scaffolding. \pb

\subsection*{Scaffolding} %{{{2

Scaffolding can be described as the process of moving from a series of
disconnected contigs to complete continuous sequence of all the contigs joined
together. The scaffolding process may not be completed but may still resolve
and join some contigs together or add additional regions of sequence to fill
gaps. In the case of a microbial genome, the process of scaffolding may
include the following steps:

\subsubsection*{Contig Orientation} %{{{3

Sequence reads may be produced from either DNA strand during the sequencing
process. Therefore the resulting contigs produced from these sequences may are
likely to be representative from either of the genome strands. Contig
orientation involves correcting the sequence data so that all contigs
represent sequence data from the same DNA strand. In the case of microbes this
orientation is often the 5`-3` strand from the origin of DNA replication.

\subsubsection*{Contig Ordering} %{{{3

The source location of each contig from the genome is unknown. Contig ordering
represents determining the order of contigs which most accurately represent
the actual genome sequence. This process can also determine if any contigs
represent extra-genomic DNA such as plasmids and therefore should be separated
from the genomic contigs. Contigs are usually ordered in reference to the
origin of replication where the contig enclosing this region is the first
contigs and subsequent contigs are labelled in the direction of DNA
replication.

\subsubsection*{Contig Distancing} %{{{3

Once the order of contigs in the genome sequence is known it is useful to
determine their approximate distance from each other. This allows the size of
the gaps between contigs to be determined. Filling in these regions in with
unknown nucleotide characters allows a draft continuous representation of the
genome to determined.

\subsubsection*{Gap Closing and Filling} %{{{3

The final stage of producing a genome sequence is to close and fill gaps in
the sequence. This may require returning to the laboratory to perform
additional sequencing to close gaps or \emph{in silico} methods may be able to
close gaps using existing data. Once all gaps in a scaffold have been closed
the genome may be considered finished. Though additional sequencing may be
performed to `proof' any low coverage regions for erroneous nucleotide
insertions, deletions or substitutions.

\subsection*{Computational Methods for Scaffolding} %{{{2

Scaffolding may be considered a combination of \emph{in vitro} and \emph{in
silico} methods. Several steps may be performed computationally by using
available data to generate a scaffold. Examples may be the use of pair reads
sequencing, where reads a known distance apart, or from comparing contigs to
a reference genome. In the case where \emph{in silico} methods are unable to
complete the scaffold in may be necessary to use PCR to amplify then use
traditional Sanger sequencing to find the sequence and close gaps. Obviously
\emph{in silico} methods are preferable as they less costly in laboratory time
and materials \cite{nagarajan2010}. \pb

For these reason may examples exist for software to generate genome scaffolds
from assembly data. These software use a variety of approaches. Synteny
approaches use a reference genome to order and orient contigs
\cite{richter2007,zhao2008}. Repeat regions which may be responsible for many
of the gaps between contigs may be reassembled using algorithms specifically
for repeat resolution \cite{mulyukov2002,koren2010}. Uniquely matching reads
may be used to try and bridge gaps between scaffolds \cite{tsai2010}. Whole
genome may use heuristic graph algorithms to build a path between contigs using
multiple input data \cite{pop2004,dayarian2010}. Finally when the scaffold
cannot be completely resolved \emph{in silico} software packages exist to
suggest the likely primers necessary for determining the sequence in gap
regions \cite{gordon2001,nagarajan2010}. Taken together these tools provide
a wide variety of possible options for scaffolding a genome should an assembler
be unable to produce a continuous sequence from the sequencing reads. \pb

Scaffolding however will inevitably require manual editing if computational
methods do not completely resolve the final genome sequence. This involves
moving an joining larges blocks of nucleotide text together in what the genome
finisher perceives to be the correct scaffold. This process may be expected to
be error-prone. Further manual editing cannot be reliably reproduced which
precludes reproducibility in a genome scaffolding process. \pb

\subsection*{The Scaffolder Software} %{{{2

The software outlined here ``Scaffolder'' is aimed at joining contigs together
to build a genome scaffold. Scaffolder uses a specific data syntax to define
how contigs are joined together. Furthermore scaffolder allows inserts into
contigs and unknown regions in a scaffold to be defined. This syntax allows
a scaffold to be edited in much easier way than editing large blocks of
nucleotides. Further more this makes genome scaffolds reproducible.

\section*{Implementation} %{{{1

  Text for this section \ldots
% }}}
% {{{ Results and Discussion
 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% Results and Discussion %%
%%
\section*{Results and Discussion}
  \subsection*{Results sub-heading}
    \subsubsection*{This is a sub-sub-heading}
      Sub-sub-sub-headings are made with the \textsl{\\subsubsection} command. \pb
      pb at end of lines ensures correct paragraph spacing.\pb
	  Text for this sub-sub-section \ldots
    \subsubsection*{Another sub-sub-sub-heading}
      Text for this sub-sub-section \ldots

  \subsection*{Another results sub-heading}
    Text for this sub-section \ldots

  \subsection*{Yet another results sub-heading}
    Text for this sub-section.  More results \ldots
% }}}
% {{{ Conclusions
%%%%%%%%%%%%%%%%%%%%%%
\section*{Conclusions}
  Text for this section \ldots
% }}}
% {{{ Availability and Requirements
\section*{Availability and Requirements}

  \begin{description}
    \item[Project name:] Scaffolder
    \item[Project home page:] \scaffolder
    \item[Operating system:] Platform Independent. Tested on Mac OS X and
    Debian.
    \item[Programming language:] Ruby
    \item[Other requirements:] RubyGems
    \item[License:] MIT
    \item[Any restrictions to use by non-academics:] None
  \end{description}

% }}}
% {{{ Competing Interests
\section*{Authors contributions}
  The authors declare no competing interests.
% }}}
% {{{ Author Contributions
\section*{Authors contributions}

MDB developed and maintains the scaffolder tool. MDB and HAB wrote the
manuscript.

% }}}
% {{{ Author's Information
\section*{Authors contributions}
    Text for this section \ldots
% }}}
% {{{ Acknowledgements
\section*{Acknowledgements}
  \ifthenelse{\boolean{publ}}{\small}{}
  Text for this section \ldots
%}}}
% {{{ Bibliography
{\ifthenelse{\boolean{publ}}{\footnotesize}{\small}
 \bibliographystyle{bmc_article}  % Style BST file
  \bibliography{article} }     % Bibliography file (usually '*.bib' ) 
%}}}
% {{{ Figures
\ifthenelse{\boolean{publ}}{\end{multicols}}{}

\section*{Figures}
  \subsection*{Figure 1 - Sample figure title}
      A short description of the figure content
      should go here.

  \subsection*{Figure 2 - Sample figure title}
      Figure legend text.
% }}}
% {{{ Tables
\section*{Tables}
  \subsection*{Table 1 - Sample table title}
    Here is an example of a \emph{small} table in \LaTeX\ using  
    \verb|\tabular{...}|. This is where the description of the table 
    should go. \par \mbox{}
    \par
    \mbox{
      \begin{tabular}{|c|c|c|}
        \hline \multicolumn{3}{|c|}{My Table}\\ \hline
        A1 & B2  & C3 \\ \hline
        A2 & ... & .. \\ \hline
        A3 & ..  & .  \\ \hline
      \end{tabular}
      }
  \subsection*{Table 2 - Sample table title}
    Large tables are attached as separate files but should
    still be described here.
% }}}
% {{{ Additional Files
\section*{Additional Files}
  \subsection*{Additional file 1 --- Sample additional file title}
    Additional file descriptions text (including details of how to
    view the file, if it is in a non-standard format or the file extension).  This might
    refer to a multi-page table or a figure.

  \subsection*{Additional file 2 --- Sample additional file title}
    Additional file descriptions text.


\end{bmcformat}
\end{document}
% }}}
