% {{{ Preamble


%% BioMed_Central_Tex_Template_v1.05
%%                                      %
%  bmc_article.tex            ver: 1.05 %
%                                       %


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                     %%
%%  LaTeX template for BioMed Central  %%
%%     journal article submissions     %%
%%                                     %%
%%         <27 January 2006>           %%
%%                                     %%
%%                                     %%
%% Uses:                               %%
%% cite.sty, url.sty, bmc_article.cls  %%
%% ifthen.sty. multicol.sty		       %%
%%									   %%
%%                                     %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                                                 %%	
%% For instructions on how to fill out this Tex template           %%
%% document please refer to Readme.pdf and the instructions for    %%
%% authors page on the biomed central website                      %%
%% http://www.biomedcentral.com/info/authors/                      %%
%%                                                                 %%
%% Please do not use \input{...} to include other tex files.       %%
%% Submit your LaTeX manuscript as one .tex document.              %%
%%                                                                 %%
%% All additional figures and files should be attached             %%
%% separately and not embedded in the \TeX\ document itself.       %%
%%                                                                 %%
%% BioMed Central currently use the MikTex distribution of         %%
%% TeX for Windows) of TeX and LaTeX.  This is available from      %%
%% http://www.miktex.org                                           %%
%%                                                                 %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


\NeedsTeXFormat{LaTeX2e}[1995/12/01]
\documentclass[10pt]{bmc_article}    



% Load packages
\usepackage{cite} % Make references as [1-4], not [1,2,3,4]
\usepackage{url}  % Formatting web addresses  
\usepackage{ifthen}  % Conditional 
\usepackage{multicol}   %Columns
\usepackage[utf8]{inputenc} %unicode support
%\usepackage[applemac]{inputenc} %applemac support if unicode package fails
%\usepackage[latin1]{inputenc} %UNIX support if unicode package fails
\urlstyle{rm}
 
 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%	
%%                                             %%
%%  If you wish to display your graphics for   %%
%%  your own use using includegraphic or       %%
%%  includegraphics, then comment out the      %%
%%  following two lines of code.               %%   
%%  NB: These line *must* be included when     %%
%%  submitting to BMC.                         %% 
%%  All figure files must be submitted as      %%
%%  separate graphics through the BMC          %%
%%  submission process, not included in the    %% 
%%  submitted article.                         %% 
%%                                             %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%                     

\def\includegraphic{}
\def\includegraphics{}

\setlength{\topmargin}{0.0cm}
\setlength{\textheight}{21.5cm}
\setlength{\oddsidemargin}{0cm} 
\setlength{\textwidth}{16.5cm}
\setlength{\columnsep}{0.6cm}

\newboolean{publ}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                              %%
%% You may change the following style settings  %%
%% Should you wish to format your article       %%
%% in a publication style for printing out and  %%
%% sharing with colleagues, but ensure that     %%
%% before submitting to BMC that the style is   %%
%% returned to the Review style setting.        %%
%%                                              %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%Review style settings
\newenvironment{bmcformat}{\begin{raggedright}\baselineskip20pt\sloppy\setboolean{publ}{false}}{\end{raggedright}\baselineskip20pt\sloppy}

%Publication style settings
%\newenvironment{bmcformat}{\fussy\setboolean{publ}{true}}{\fussy}

% Scaffolder url
\urldef{\scaffolder}\url{http://www.michaelbarton.me.uk/projects/scaffolder/}

% Begin ...
\begin{document}
\begin{bmcformat}
% }}}
% {{{ Title

\title{Scaffolder - Software for Microbial Genome Scaffolding.}

\author{
  Michael D Barton$^{1}$%
  \email{Michael D Barton - mail@michaelbarton.me.uk}%
\and
  Hazel A Barton\correspondingauthor$^1$%
  \email{Hazel A Barton\correspondingauthor - bartonh@nku.edu}%
      }

\address{\iid(1) Department of Biological Sciences, Northern Kentucky%
University, Nunn Drive, Highland Heights, KY 41076 }%

\maketitle
%}}}
% {{{ Abstract

\begin{abstract}

  \paragraph*{Background:} Assembly of short read sequencing data can result in
  a fragmented series of contigs. Therefore a common step in a genome project
  is to join neighbouring sequence regions together and fill gaps using insert
  data generated through PCR. This scaffolding step, however, is non-trivial
  and requires manually editing large blocks of nucleotide sequence. Joining
  sequence contigs together also hides the source of each region in the final
  genome sequence. Taken together, these considerations may make reproducing or
  editing an existing genome build difficult.

  \paragraph*{Methods:} The software outlined "Scaffolder" is implemented in
  the Ruby programming language and can be installed via the RubyGems software
  management system. Genome scaffolds are defined using YAML - a markup
  language, which is both human and machine-readable. Command line binaries and
  extensive documentation are provided.

  \paragraph*{Results:} This software allows a genome build to be defined in
  terms of the constituent contigs using a relatively simple to write and edit
  scaffold file syntax. The Scaffolder syntax further allows inserts to be
  added to the scaffold for the purpose of filling unknown sequence regions.
  Defining the genome construction in a file makes the scaffolding process
  reproducible and comparatively easier to edit than raw fasta.

  \paragraph*{Conclusions:} Scaffolder is easy to use genome scaffolding
  software. This tool promotes reproducibility and maintenance in building
  a genome. Scaffolder can be found at \scaffolder.

\end{abstract}

\ifthenelse{\boolean{publ}}{\begin{multicols}{2}}{}

% }}}
% {{{ Background
\section*{Background}
 Text for this section.
% }}}
\section*{Implementation} %{{{1

\subsection*{Code and Dependencies} %{{{2

Scaffolder is written in the Ruby programming language, version 1.8.7
\cite{ruby-lang}. Unit tests were implemented during development to maintain
functionality of individual elements of source code. Integration tests were
also written to test the scaffold software interface as a whole. These unit and
integration tests were written using the RSpec and Cucumber libraries
respectively\cite{rspec}.\pb

The Scaffolder source code is documented using the Yard library\cite{yard}.
Unix manual pages for the command line interface were generated using the Ronn
library\cite{ronn}. The manipulation of biological sequences in Scaffolder is
performed using the BioRuby library\cite{bioruby}. A full list of the software
dependencies in scaffolder can be found in the Gemfile in the root of source
code directory.\pb

\subsection*{Scaffold file syntax} %{{{2

The specification of the nucleotide sequences comprising the genome scaffold is
written using YAML \cite{yaml}. YAML is a data format using whitespace and
identation to produce a standardised machine readable structure. As a commonly
accepted data format third-party developers have the option to generate
a genome scaffold using their programming language of choice. As a standardised
format YAML formatted scaffold files can checked for syntax validity using
third party tools such as Kwalify\cite{kwalify}. \pb

In addition to being machine readable it also important that a genome scaffold
file also be easy to edit manually. As such this was the second reason the YAML
syntax was selected for the scaffold file data format. YAML has a concise and
compact syntax and when compared with formats such as extensible markup
language (XML) which require angular braces (`$<$' and `$>$') and open and
closing tags to delimite each data field.

By choosing a human readable syntax it is aimed that the scaffold should be
easier to edit and change manually in simple text edit.  A scenario which we
anticipate is common in the on-going process of scaffolding a genome sequence.
\pb

The entries in the YAML scaffold are defined as a list. Each entry in the list
corresponds to a region sequence in the final output sequence which follows the
previous entry in the scaffold and precedes the following entry.  Each
attribute defined for each entry in the list allows specifications for the
entry such as if any starting sequence should be clipped from the beginning or
end, or if the sequence should be reversed. \pb

INSERT IMAGE OF SCAFFOLD GRAPH AND RESULTING OUTPUT.

\subsection*{Parsing the YAML scaffold} %{{{2

Each stretch of sequence in the genome scaffold is defined as a list entry in
the YAML scaffold file. When parsing the scaffold file Scaffolder iterates over
each scaffold entry to create an instance of the Scaffolder::Region class.\pb

The Region class is inherited by all types of sequence in the scaffold. The
Region provides helper methods common to the tasks in manipulating scaffold
sequence, subsequencing and reversing. In particular the region class has
a class annotation method\cite{class-annotations} which allows attributes in
the scaffold method to be defined in the class. This negates having to write
a custom method to deal with each scaffold variable.  Furthermore this makes
the process of transferring attributes from the scaffold file to each region
instance much simpler - each yaml attribute corresponds to a method call on the
instantiated class. The scaffold file effectively becomes a DSL for populating
the scaffold region instances while still maintaining the machine readable data
qualities of YAML.\pb

The region class provides a sequence hook method. This method returns by
default the sequence corresponding to the fasta sequence defined by the source
tag which is then sublclassed or reversed if necessary. Subclasses of the
region class should override this method to provide custom functionality in
manipulating the source sequence. For example, the Scaffolder::Region::Sequence
class overrides the sequence method add any sequence inserts specified in the
scaffold file.\pb

\section*{Results and Discussion} %{{{1
  \subsection*{Results sub-heading}
    \subsubsection*{This is a sub-sub-heading}
      Sub-sub-sub-headings are made with the \textsl{\\subsubsection} command. \pb
      pb at end of lines ensures correct paragraph spacing.\pb
	  Text for this sub-sub-section \ldots
    \subsubsection*{Another sub-sub-sub-heading}
      Text for this sub-sub-section \ldots

  \subsection*{Another results sub-heading}
    Text for this sub-section \ldots

  \subsection*{Yet another results sub-heading}
    Text for this sub-section.  More results \ldots
% }}}
% {{{ Conclusions
%%%%%%%%%%%%%%%%%%%%%%
\section*{Conclusions}
  Text for this section \ldots
% }}}
% {{{ Availability and Requirements
\section*{Availability and Requirements}

  \begin{description}
    \item[Project name:] Scaffolder
    \item[Project home page:] \scaffolder
    \item[Operating system:] Platform Independent. Tested on Mac OS X and
    Debian.
    \item[Programming language:] Ruby
    \item[Other requirements:] RubyGems
    \item[License:] MIT
    \item[Any restrictions to use by non-academics:] None
  \end{description}

% }}}
% {{{ Competing Interests
\section*{Authors contributions}
  The authors declare no competing interests.
% }}}
% {{{ Availability and Requirements
\section*{List of Abbreviations Used}

  \begin{description}
    \item[YAML:] YAML Ain't Markup Language\cite{yaml}
  \end{description}

% }}}
% {{{ Author Contributions
\section*{Authors contributions}

MDB developed and maintains the scaffolder tool. MDB and HAB wrote the
manuscript.

% }}}
% {{{ Author's Information
\section*{Authors contributions}
    Text for this section \ldots
% }}}
% {{{ Acknowledgements
\section*{Acknowledgements}
  \ifthenelse{\boolean{publ}}{\small}{}
  Text for this section \ldots
%}}}
% {{{ Bibliography
{\ifthenelse{\boolean{publ}}{\footnotesize}{\small}
 \bibliographystyle{bmc_article}  % Style BST file
  \bibliography{article} }     % Bibliography file (usually '*.bib' ) 
%}}}
% {{{ Figures
\ifthenelse{\boolean{publ}}{\end{multicols}}{}

\section*{Figures}
  \subsection*{Figure 1 - Sample figure title}
      A short description of the figure content
      should go here.

  \subsection*{Figure 2 - Sample figure title}
      Figure legend text.
% }}}
% {{{ Tables
\section*{Tables}
  \subsection*{Table 1 - Sample table title}
    Here is an example of a \emph{small} table in \LaTeX\ using  
    \verb|\tabular{...}|. This is where the description of the table 
    should go. \par \mbox{}
    \par
    \mbox{
      \begin{tabular}{|c|c|c|}
        \hline \multicolumn{3}{|c|}{My Table}\\ \hline
        A1 & B2  & C3 \\ \hline
        A2 & ... & .. \\ \hline
        A3 & ..  & .  \\ \hline
      \end{tabular}
      }
  \subsection*{Table 2 - Sample table title}
    Large tables are attached as separate files but should
    still be described here.
% }}}
% {{{ Additional Files
\section*{Additional Files}
  \subsection*{Additional file 1 --- Sample additional file title}
    Additional file descriptions text (including details of how to
    view the file, if it is in a non-standard format or the file extension).  This might
    refer to a multi-page table or a figure.

  \subsection*{Additional file 2 --- Sample additional file title}
    Additional file descriptions text.


\end{bmcformat}
\end{document}
% }}}
