% {{{ Preamble


%% BioMed_Central_Tex_Template_v1.05
%%                                      %
%  bmc_article.tex            ver: 1.05 %
%                                       %


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                     %%
%%  LaTeX template for BioMed Central  %%
%%     journal article submissions     %%
%%                                     %%
%%         <27 January 2006>           %%
%%                                     %%
%%                                     %%
%% Uses:                               %%
%% cite.sty, url.sty, bmc_article.cls  %%
%% ifthen.sty. multicol.sty		       %%
%%									   %%
%%                                     %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                                                 %%	
%% For instructions on how to fill out this Tex template           %%
%% document please refer to Readme.pdf and the instructions for    %%
%% authors page on the biomed central website                      %%
%% http://www.biomedcentral.com/info/authors/                      %%
%%                                                                 %%
%% Please do not use \input{...} to include other tex files.       %%
%% Submit your LaTeX manuscript as one .tex document.              %%
%%                                                                 %%
%% All additional figures and files should be attached             %%
%% separately and not embedded in the \TeX\ document itself.       %%
%%                                                                 %%
%% BioMed Central currently use the MikTex distribution of         %%
%% TeX for Windows) of TeX and LaTeX.  This is available from      %%
%% http://www.miktex.org                                           %%
%%                                                                 %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


\NeedsTeXFormat{LaTeX2e}[1995/12/01]
\documentclass[10pt]{bmc_article}    



% Load packages
\usepackage{cite} % Make references as [1-4], not [1,2,3,4]
\usepackage{url}  % Formatting web addresses  
\usepackage{ifthen}  % Conditional 
\usepackage{multicol}   %Columns
\usepackage[utf8]{inputenc} %unicode support
%\usepackage[applemac]{inputenc} %applemac support if unicode package fails
%\usepackage[latin1]{inputenc} %UNIX support if unicode package fails
\urlstyle{rm}
 
 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%	
%%                                             %%
%%  If you wish to display your graphics for   %%
%%  your own use using includegraphic or       %%
%%  includegraphics, then comment out the      %%
%%  following two lines of code.               %%   
%%  NB: These line *must* be included when     %%
%%  submitting to BMC.                         %% 
%%  All figure files must be submitted as      %%
%%  separate graphics through the BMC          %%
%%  submission process, not included in the    %% 
%%  submitted article.                         %% 
%%                                             %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%                     

\def\includegraphic{}
\def\includegraphics{}

\setlength{\topmargin}{0.0cm}
\setlength{\textheight}{21.5cm}
\setlength{\oddsidemargin}{0cm} 
\setlength{\textwidth}{16.5cm}
\setlength{\columnsep}{0.6cm}

\newboolean{publ}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                              %%
%% You may change the following style settings  %%
%% Should you wish to format your article       %%
%% in a publication style for printing out and  %%
%% sharing with colleagues, but ensure that     %%
%% before submitting to BMC that the style is   %%
%% returned to the Review style setting.        %%
%%                                              %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%Review style settings
\newenvironment{bmcformat}{\begin{raggedright}\baselineskip20pt\sloppy\setboolean{publ}{false}}{\end{raggedright}\baselineskip20pt\sloppy}

%Publication style settings
%\newenvironment{bmcformat}{\fussy\setboolean{publ}{true}}{\fussy}

% Scaffolder url
\urldef{\scaffolder}\url{http://www.michaelbarton.me.uk/projects/scaffolder/}

% Begin ...
\begin{document}
\begin{bmcformat}
% }}}
% {{{ Title

\title{Scaffolder - Software for Microbial Genome Scaffolding.}

\author{
  Michael D Barton$^{1}$%
  \email{Michael D Barton - mail@michaelbarton.me.uk}%
\and
  Hazel A Barton\correspondingauthor$^1$%
  \email{Hazel A Barton\correspondingauthor - bartonh@nku.edu}%
      }

\address{\iid(1) Department of Biological Sciences, Northern Kentucky%
University, Nunn Drive, Highland Heights, KY 41076 }%

\maketitle
%}}}
% {{{ Abstract

\begin{abstract}

  \paragraph*{Background:} Assembly of short read sequencing data can result in
  a fragmented series of contigs. Therefore a common step in a genome project
  is to join neighbouring sequence regions together and fill gaps using insert
  data generated through PCR. This scaffolding step, however, is non-trivial
  and requires manually editing large blocks of nucleotide sequence. Joining
  sequence contigs together also hides the source of each region in the final
  genome sequence. Taken together, these considerations may make reproducing or
  editing an existing genome build difficult.

  \paragraph*{Methods:} The software outlined "Scaffolder" is implemented in
  the Ruby programming language and can be installed via the RubyGems software
  management system. Genome scaffolds are defined using YAML - a markup
  language, which is both human and machine-readable. Command line binaries and
  extensive documentation are provided.

  \paragraph*{Results:} This software allows a genome build to be defined in
  terms of the constituent contigs using a relatively simple to write and edit
  scaffold file syntax. The Scaffolder syntax further allows inserts to be
  added to the scaffold for the purpose of filling unknown sequence regions.
  Defining the genome construction in a file makes the scaffolding process
  reproducible and comparatively easier to edit than raw fasta.

  \paragraph*{Conclusions:} Scaffolder is easy to use genome scaffolding
  software. This tool promotes reproducibility and maintenance in building
  a genome. Scaffolder can be found at \scaffolder.

\end{abstract}

\ifthenelse{\boolean{publ}}{\begin{multicols}{2}}{}

% }}}
% {{{ Background
\section*{Background}
 Text for this section.\cite{koon,oreg,khar,zvai,xjon,schn,pond,smith,marg,hunn,advi,koha,mouse}
% }}}
% {{{ Implementation
%%%%%%%%%%%%%%%%%%%%%%
\section*{Implementation}
  Text for this section \ldots
% }}}
\section*{Results and Discussion} %{{{1

\subsection*{Scaffolder Description} %{{{2

Individual reads produced from nucleotide sequencing can be assembled into
contiguous regions (``contigs'') of many overlapping individual reads. These
reads may however not represent a complete genome sequence where all contigs do
not result in a single complete sequence. The software outlined here
``Scaffolder'' allows these contigs to be further joined together into
a continuous single sequence. This selection, order and direction of contigs
into a final draft genome is often described as the `scaffold'. Scaffolder
allows the manual or computational specification of a genome scaffold by
specifying which contig sequences are used in simple to edit configuration
files. \pb

The scaffold file specifies how contigs should be joined together in the draft
genome sequence. The raw nucleotide sequence for each contig is stored in
a fasta file, as might be produced from assembly of sequencing reads using
programs such as Newbler\cite{miller2010} or ABySS\cite{simpson2009}.
Scaffolder then the contig to be selected by specifying the name of the contig
in the scaffold file. The scaffold file also allows a sub-sequence or the
reverse complement of a contig to used. This therefore preserves the original
sequence data but also allowing sequences to manipulated in the final genome
build. \pb

The majority of a genome can be estimated from scaffolding assembled contigs
into the most parsimonious build. There may however still be internal gaps in
the contigs from mate pair assembly based on estimated mate pair insert size.
Scaffolder allows inserts to be added to scaffolded so that gaps may be bridged
and filled - a common process in finishing a genome sequence. Inserts may be
treated the same way as a contig where the raw sequence, as may be produced by
PCR is stored in a fasta file, and then trimmed and/or reversed to fill the gap
in the contig. \pb

Scaffolder allows the specification of unknown regions in the scaffold. These
are areas of the genome where the nucleotide sequence is unknown but the
approximate size region may be estimated from mate pair insert size or from
mapping contigs to reference genome and estimating the size of the gap. These
unknown regions allow a continuous draft sequence to be scaffolded but whilst
still highlighting regions waiting to be resolved. \pb

\subsection*{Defining a scaffold in a text file} %{{{2

The genome scaffold is wriiten in YAML format. YAML is a data format designed
to be both machine readable and easy to manually edit. An example of this
format illustrating all the features of scaffolder is shown in Figure 1. \pb

\subsubsection*{Simple sequence region} %{{{3

The first line of the file begins with three dashes. This is required to
signify the start of a YAML document. Lines 2 begins with a dash character `-'
which denotes the start of an entry in the sequence scaffold, in this case the
first entry. Line 3 is indented by two spaces which therefore means this line
contains an entry type. The ``sequence'' tag is used indicate a sequence is
inserted at this region in the scaffold. The following line contain is
indented by two spaces and indicates that this an attribute of the sequence
entry. In this case the ``source'' keyword is used to indicate used which
sequence should be inserted into the scaffold. This sequence is indentified by
the first word of the fasta header of the selected sequence. This indicates
a simple sequence region added to the scaffold. \pb

\subsubsection*{Unresolved seqeunce region} %{{{3

The following line has no indentation and begins with dash indicating the next
entry in the scaffold. The following entry is indentified by the ``unresolved''
tag. This identifies a region of unknown nucleotide sequence but is included
because the approximate size of the region is known. The following line defines
that the size of this unknown region in the genome scaffold should be 20
nucleotides. This therefore results in the region being filled with 20 `N'
characters in draft sequence. \pb

\subsubsection*{Trimmed sequence region with multiple inserts} %{{{3

The following entry is another sequence using the sequence from 'sequence2' in
the scaffold file. This sequence entry has three further attributes in addition
to defining the source sequence. The first two are start and stop. These are
used to trim the final sequence inserted into the scaffold. In this case the
first two nucleotides of the sequence are removed so that that sequence starts
at position 3 of the original raw sequence. Similarly the sequence after
position is trimmed away after position 100. These sequence is also reverse
complemented using the ``reverse'' tag. Reverse complementation is performed
after the sequence has been trimmed to the specified coordinates. \pb

The final attribute for this sequence entry defines inserts to be added to the
sequence. These are additional regions of sequence that can be added to
a sequence entry to alter the existing sequence such to fill gaps or correct
sequencing errors. Inserts are defined using the ``inserts'' tag. Each insert
is then listed similarly to the entries in the scaffold: beginning with a dash
on one line then indented by an additional two spaces on the following lines.
\pb

Inserts are listed similarly to sequence entries where reverse, start and stop
attributes can be used to trim and reverse the sequence. The `source' tag is
used to indentify which fasta sequence entry should be used as the insert.  The
insert has two additional tags ``open'' and ``close'' which determine where the
insert is added to the hosting sequence. These coordinates are inclusive
replaceing the coordinate nucleotides. The open and close positions need not be
the same size as the insert size. Any difference in size to the replaced
nucleotides will result in the length on the encoding sequence being adjusted.
\pb

The next insert entry shows an insert with only the open position specified.
This is to illustrate that only the one of the `open' or `close' keywords is
sufficient to add an insert. If only one of these words is used then the
opposing insert position is calculated from the length on the insert sequence
in the fasta file. This feature is designed for using inserts to close gaps
where the length of the insert may not completely fill the gap but can still be
used to fill a portion of the gap sequence. \pb

\subsubsection*{Scaffolder file attribute processing} %{{{3

Each scaffolder sequence entry can have multiple attributes which applied in
different order can result in a different sequence. For instance reverse
complementing the sequence then adding inserts will have a different effect
compared with performing these actions in the reverse order. Scaffolder
therefore follows the following protocol for processing each entry in the
scaffold file. \pb

\paragraph{Reverse sort inserts:} If the entry is a sequence and has sequence
inserts these are reverse sorted by insert close position. The insert with the
largest close position is first and and the smallest close position is last.
This is done to preserve the the insert open/close coordinates as they are
added to the scaffold. \pb

\paragraph{Update sequence with insert:} Each insert is added to the sequence
replacing the region specified by the open an close position with the specified
source sequence.\pb

\paragraph{Update sequence start/stop co-ordinates:} If the insert added to the
sequence is a different length than the region replaced the start and stop
co-ordinates of the sequence are correspondingly updated if necessary. E.g. if
a 7 bp insert filled a bp region this would result in a 2bp difference in the
length of the sequence. The stop coordinated would be updated to reflect this.
\pb

\paragraph{Sub-sequence selected:} If the sequence is has start or stop
sequences specified these are used to trip the sequence to these co-ordinates.
\pb 

\paragraph{Sub-sequence reverse complemented: } If specified with the reverse keyworkd the sequence is reverse complemented.
\pb

\subsection*{Scaffolder usage} %{{{2

Scaffolder provides a ruby API that allows the scaffold and designated
sequences to be manipulated through a ruby programme. This allows scaffolder to
be integrated into exiting genomics works flows such as using Rake.  Scaffolder
all provides a command line binary for the manipulation of the scaffold. This
binary behaves as a standard Unix command line interface returning appropriate
exit statuses and providing manual pages for each of the commands. These two
methods are outlined in the following three sections. \pb

\subsubsection*{Installation} %{{{3

Scaffolder relies on both the ruby programming language and the RubyGems
package management system. These are commonly installed on recent versions of
Mac OSX and linux distrubtions. The installation of these can be tested on the
command line by typing the following commands. \pb

\begin{verbatim}
ruby -v
gem -v
\end{verbatim}

If both these commands return without error this indicates the necessary
requirements to install scaffolder. The scaffolder API and command line tools
can then be installed with the follow command.

\begin{verbatim}
gem install scaffolder scaffolder-tools
\end{verbatim}

Finally to ensure that the command line tools are accessible these the
RubyGems binary directory should be in the command line path. This can be done
by adding the following to the ~/.bashrc file or other corresponding shell
configure file. Assuming that ~/.gem is the install path for RubyGems.

\begin{verbatim}
export PATH=~/.gem/bin:$PATH
\end{verbatim}

\subsubsection*{Command line interface} %{{{3

Scaffolder provides a command line interface to validate and build a genome
sequence from a scaffold. Assuming scaffolder has been installed and the
binary is available in the command line path the following command can be used
to invoke scaffolder. This will list the available commands for running
scaffolder. \pb

\begin{verbatim}
$ scaffolder
\end{verbatim}

Help on each command can be found by type `help' followed by the name of the
command. This will return the Unix manual page for the command of interest.
For example the following command will return the manual page for sequence
listing the available options for running the sequence command. \pb

\begin{verbatim}
$ scaffolder help sequence
\end{verbatim} 

To the generate the fasta sequence from a scaffold the sequence command should
be used. This should be passed the name of the file with containing the YAML
formatted scaffold and the corresponding fasta file of sequences. This will
then return the complete drafted sequence to the standard out. \pb

\begin{verbatim}
$ scaffolder sequence scaffold.yml sequences.fna
\end{verbatim}

Scaffolder also provides a function to test the scaffold for overlapping
insert sequence co-ordinates. As described in the section above overlapping
inserts will result in unexpected behaviour as the coordinates for the first
insert will change the expected insert location for the second insert. The
validate command will check the scaffold for any overlapping inserts and
return them as a YAML formatted list describing each one. \pb

\begin{verbatim}
$ scaffolder validate scaffold.yml sequences.fna
\end{verbatim}

\subsubsection*{Application programming interface} %{{{3

The scaffolder tool has a defined application programming interface (API) to
allow ruby programs to manipulate the scaffold. Detailing the complete API
would be overwhelming to this document. The scaffolder API is fully documented
online using at the scaffolder website. The following code snippet gives and
outline of how the API may be used to create a script to replicate the
function of the `scaffolder sequence' command described above. \pb

\begin{verbatim}
  #!/usr/bin/env ruby

  require 'rubygems'
  require 'scaffolder'

  scaffold_file = ARGV.shift
  sequence_file = ARGV.shift

  scaffold = Scaffolder.new(
    YAML.load(File.read(scaffold_file)),
    sequence_file)

  sequence = scaffold.map{|entry| entry.sequence }.join

  puts sequence
\end{verbatim}

This program illustrates that the scaffold object is initialised with two arguments: the YAML loaded scaffold file and the location of the fasta formatted sequence file. The scaffold is then enumerable as the next line illustrates the trivial code required to generate the scaffold. Further details of the attributes for each entry in the scaffold can be found in the documentation for the Scaffolder::Region class. \pb

\subsection*{Discussion of scaffolder approach} %{{{2

The main aims of scaffolder are two fold. The first is usuability and the second is reproducibility.

\subsubsection{Simple to use}

Scaffolder is designed to be as simple to use as possible. Assuming the
existence of Ruby and RubyGems, scaffolder can be installed with a single
command. Scaffolder uses a minimal and compact syntax to described the
scaffold. This syntax is simple to construct and edit whilst being succint and
readable in describing how the scaffold was constructed. Scaffolder provides
extensive documentation both for the command line tools and for the Scaffolder
Ruby API.

\subsubsection{Reproducible genome construction}

Constructing a genome sequence using a plain text scaffold file means the same
sequence is always output for the same file. The sequence produced from
manually joining separate contigs in a text editor can not be reliably
reproduced. Constructing the scaffold in a plain text file means that the
sequence is reproducible. Furthermore the source of each region in final
sequence can be identified. The scaffold file provides a readable record of
how the sequence was constructed. The scaffold file also means the scaffold is
easer to edit once already constructed. Much more difficult to do when working
on the sequence alone. The YAML text format is also much easier to compare
differences using standard diff commands. This makes scaffold files much more
simple to store in version control and to see the stages in the construction
of the genome sequence.

\section*{Conclusions} %{{{1

Scaffolder follows the Unix philospophy ``Do one thing and do it well''.
Scaffolder focuses on creating a simple to use and install tool for generating
a genome scaffold. The YAML data format for creating a scaffold file is
standardised and easily manipulated programmatically. This thereby follows
a second Unix tenet: ``If your data structures are good enough, the algorithm
to manipulate them should be trivial.''

\section*{Availability and Requirements} %{{{1

  \begin{description}
    \item[Project name:] Scaffolder
    \item[Project home page:] \scaffolder
    \item[Operating system:] Platform Independent. Tested on Mac OS X and
    Debian.
    \item[Programming language:] Ruby
    \item[Other requirements:] RubyGems
    \item[License:] MIT
    \item[Any restrictions to use by non-academics:] None
  \end{description}

% }}}
% {{{ Competing Interests
\section*{Authors contributions}
  The authors declare no competing interests.
% }}}
% {{{ Author Contributions
\section*{Authors contributions}

MDB developed and maintains the scaffolder tool. MDB and HAB wrote the
manuscript.

% }}}
% {{{ Author's Information
\section*{Authors contributions}
    Text for this section \ldots
% }}}
% {{{ Acknowledgements
\section*{Acknowledgements}
  \ifthenelse{\boolean{publ}}{\small}{}
  Text for this section \ldots
%{{{1 Bibliography
{\ifthenelse{\boolean{publ}}{\footnotesize}{\small}
 \bibliographystyle{bmc_article}  % Style BST file
  \bibliography{article} }     % Bibliography file (usually '*.bib' ) 

\ifthenelse{\boolean{publ}}{\end{multicols}}{}

\section*{Figures} %{{{1

  \subsection*{Figure 1 - Sample figure title}

  A short description of the figure content should go here.

\section*{Tables} %{{{1

  \subsection*{Table 1 - Sample table title}
    Here is an example of a \emph{small} table in \LaTeX\ using  
    \verb|\tabular{...}|. This is where the description of the table 
    should go. \par \mbox{}
    \par
    \mbox{
      \begin{tabular}{|c|c|c|}
        \hline \multicolumn{3}{|c|}{My Table}\\ \hline
        A1 & B2  & C3 \\ \hline
        A2 & ... & .. \\ \hline
        A3 & ..  & .  \\ \hline
      \end{tabular}
      }
  \subsection*{Table 2 - Sample table title}
    Large tables are attached as separate files but should
    still be described here.

\section*{Additional Files} %{{{1

  \subsection*{Additional file 1 --- Sample additional file title}
    Additional file descriptions text (including details of how to
    view the file, if it is in a non-standard format or the file extension).  This might
    refer to a multi-page table or a figure.

  \subsection*{Additional file 2 --- Sample additional file title}
    Additional file descriptions text.


\end{bmcformat}

\end{document}
